### PR TITLE
`DataLoader`s 8: docs, guide, etc

### PR DESCRIPTION
**Pull request by teh-cmc**
Some docs for the whole thing.

It heavily relies on linking to the examples for exact details and to avoid rot.

It's probably extremely bad. Help welcome.

---

Part of a series of PRs to make it possible to load _any_ file from the local filesystem, by any means, on web and native:
- #4516
- #4517 
- #4518 
- #4519 
- #4520 
- #4521 
- #4565
- #4566
- #4567

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Full build: [app.rerun.io](https://app.rerun.io/pr/4516/index.html)
  * Partial build: [app.rerun.io](https://app.rerun.io/pr/4516/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json) - Useful for quick testing when changes do not affect examples in any way
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4516)
- [Docs preview](https://rerun.io/preview/af8fb0d81c269a6ff8f2587e6135d927038eb966/docs) 
- [Examples preview](https://rerun.io/preview/af8fb0d81c269a6ff8f2587e6135d927038eb966/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: docs/content/getting-started/opening-any-file.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+- via the Rerun CLI (e.g. `rerun myfile.jpeg`),
+- using drag-and-drop,
+- using the open dialog in the Rerun Viewer.
+
+All these file loading methods support loading a single file, many files at once (e.g. `rerun myfiles/*`), or even folders.
+
+⚠ Drag-and-drop of folders does not yet work on the web version of the Rerun Viewer ⚠
+
+The [`DataLoader`](https://github.com/rerun-io/rerun/blob/main/crates/re_data_source/src/data_loader/mod.rs) trait takes care of loading files.
+Rerun ships with a bunch of builtin implementations for that trait, which handle the following filetypes:
```
Comment (nikolausWest) [inline docs/content/getting-started/opening-any-file.md:18]:
```suggestion
The following file types have built-in support in the Rerun Viewer:
```

File: docs/content/getting-started/opening-any-file.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+- 3D models: `gltf`, `glb`, `obj`
+- Images: `avif`, `bmp`, `dds`, `exr`, `farbfeld`, `ff`, `gif`, `hdr`, `ico`, `jpeg`, `jpg`, `pam`, `pbm`, `pgm`, `png`, `ppm`, `tga`, `tif`, `tiff`, `webp`.
+- Point clouds: `ply`.
+- Text files: `md`, `txt`.
+
+With the exception of `rrd` files that can be streamed from an HTTP URL (e.g. `rerun https://demo.rerun.io/version/latest/examples/dna/data.rrd`), we only support loading files from the local filesystem for now, with [plans to make this generic over any URI and protocol in the future](https://github.com/rerun-io/rerun/issues/4525).
+
+## Adding support for arbitrary filetypes
+
+If the builtin `DataLoader`s don't cover your needs, we expose two ways of implementing and registering your own loaders.
```
Comment (nikolausWest) [inline docs/content/getting-started/opening-any-file.md:29]:
```suggestion
Internally, the [`DataLoader`](https://github.com/rerun-io/rerun/blob/main/crates/re_data_source/src/data_loader/mod.rs) trait takes care of loading files in the viewer.

If the builtin `DataLoader`s don't cover your needs, we expose two ways of implementing and registering your own loaders.
```

File: docs/content/getting-started/opening-any-file.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+The easiest way to do so is by implementing what we call an "external loader": an executable -- written in any language that the Rerun SDK ships for -- that is available on your `$PATH`.
+
+This executable is just a vanilla Rerun logger.
+It will be called by the Rerun Viewer when the user opens a file, and be passed the path to that file.
+From there, it can log data as usual, using the [`stdout` logging sink](https://www.rerun.io/docs/reference/sdk-operating-modes).
+
+The Rerun Viewer will then automatically load the data streamed to the external loader's standard output.
+
+Check out our examples for [C++](https://github.com/rerun-io/rerun/tree/main/examples/cpp/external_data_loader), [Python](https://github.com/rerun-io/rerun/tree/main/examples/python/external_data_loader) and [Rust](https://github.com/rerun-io/rerun/tree/main/examples/rust/external_data_loader) that cover every steps in details.
+
```
Comment (nikolausWest) [inline docs/content/getting-started/opening-any-file.md:42]: I didn't really follow this section. How does Rerun know what binary to call for a given file? That needs to be explicitly spelled out in this section. I also think there should be a short page under _Reference_ that explains the binary registration + the viewer -> binary protocol (even if it's very simple). I don't believe our users are used to looking in the examples' comments for a description of a communication protocol

**Review by nikolausWest — CHANGES_REQUESTED** [review]
Good in general but I think we need to be a bit more explicit about how the binary plugin system works

File: docs/content/howto/open-any-file.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+This gives `DataLoader`s maximum flexibility to decide what files they are interested in, as opposed to e.g. only being able to look at a file's extension.
+
+Once notified, a `DataLoader` can return a [`DataLoaderError::Incompatible`](https://docs.rs/re_data_source/latest/re_data_source/enum.DataLoaderError.html#variant.Incompatible?speculative-link) error to indicate that it doesn't support a given file type.
+If, and only if, all loaders known to the Viewer return an `Incompatible` error code, then an error message is shown to the user indicating that this file type is not (_yet_) supported.
+
+In these instances of unsupported files, we expose two ways of implementing and registering your `DataLoader`s, explained below.
+
+### External data-loaders
+
+The easiest way to create your own `DataLoader` is by implementing what we call an "external loader": an executable -- written in any language that the Rerun SDK ships for -- that is available on your `$PATH`.
```
Comment (nikolausWest) [inline docs/content/howto/open-any-file.md:43]:
```suggestion
The easiest way to create your own `DataLoader` is by implementing what we call an "external loader": a stand alone executable written in any language that the Rerun SDK ships for. Any executable on your `$PATH` with a name that starts with `rerun_loader` will be treated as a `DataLoader`.
```

File: docs/content/howto/open-any-file.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+This executable is just a vanilla Rerun logger.
```
Comment (nikolausWest) [inline docs/content/howto/open-any-file.md:45]:
```suggestion
This executable takes a file path as input on `stdin` and outputs Rerun logs on `stdout`.
```

**Review by nikolausWest — APPROVED** [review]
Nice! (just a couple small suggestions)

File: docs/content/howto/open-any-file.md (hunk 4 above)
Comment (teh-cmc) [reply docs/content/howto/open-any-file.md:43]: Oof yeah how did I forget that part

**Pull request merged**